File: src/main/kotlin/texlab/LatexLanguageServer.kt
package texlab

import com.google.gson.JsonPrimitive
import kotlinx.coroutines.*
import kotlinx.coroutines.future.await
import kotlinx.coroutines.future.future
import org.eclipse.lsp4j.*
import org.eclipse.lsp4j.jsonrpc.messages.Either
import org.eclipse.lsp4j.jsonrpc.services.JsonDelegate
import org.eclipse.lsp4j.services.LanguageServer
import org.eclipse.lsp4j.services.WorkspaceService
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import texlab.build.*
import texlab.completion.CompletionProvider
import texlab.completion.bibtex.BibtexCitationActor
import texlab.completion.latex.data.LatexComponentDatabase
import texlab.definition.DefinitionProvider
import texlab.diagnostics.DiagnosticsProvider
import texlab.diagnostics.LatexLinterConfig
import texlab.folding.FoldingProvider
import texlab.formatting.BibtexFormatter
import texlab.formatting.BibtexFormatterConfig
import texlab.highlight.LatexLabelHighlightProvider
import texlab.hover.HoverProvider
import texlab.link.LatexIncludeLinkProvider
import texlab.metadata.BibtexEntryTypeMetadataProvider
import texlab.metadata.LatexComponentMetadataProvider
import texlab.metadata.MetadataProvider
import texlab.provider.FeatureProvider
import texlab.provider.FeatureRequest
import texlab.references.ReferenceProvider
import texlab.rename.RenameProvider
import texlab.resolver.InvalidTexDistributionException
import texlab.resolver.LatexResolver
import texlab.resolver.TexDistributionError
import texlab.search.ForwardSearchConfig
import texlab.search.ForwardSearchResult
import texlab.search.ForwardSearchTool
import texlab.symbol.SymbolProvider
import texlab.syntax.BibtexSyntaxTree
import texlab.syntax.LatexSyntaxTree
import texlab.syntax.bibtex.BibtexDeclarationSyntax
import java.io.File
import java.io.IOException
import java.net.URI
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.concurrent.CompletableFuture
import kotlin.coroutines.CoroutineContext
import kotlin.streams.toList

@ObsoleteCoroutinesApi
class LatexLanguageServer : LanguageServer, LatexTextDocumentService, WorkspaceService, CoroutineScope {
    override val coroutineContext: CoroutineContext = Dispatchers.Default + SupervisorJob()

    private val logger: Logger = LoggerFactory.getLogger("")
    private lateinit var client: LatexLanguageClient

    private var workspaceRootDirectory: Path? = null
    private val workspaceActor = WorkspaceActor()

    private val progressListener = object : ProgressListener {
        override fun onReportProgress(params: ProgressParams) {
            client.progress(params)
        }
    }

    private val resolver: Deferred<LatexResolver> = async(start = CoroutineStart.LAZY) {
        try {
            LatexResolver.create()
        } catch (e: InvalidTexDistributionException) {
            val message = when (e.error) {
                TexDistributionError.KPSEWHICH_NOT_FOUND ->
                    """An error occurred while executing `kpsewhich`.
                        |Please make sure that your distribution is in your PATH environment variable
                        |and provides the `kpsewhich` tool.
                    """.trimMargin()
                TexDistributionError.UNKNOWN_DISTRIBUTION ->
                    """Your TeX distribution is not supported.
                        |Please install a supported distribution.
                    """.trimMargin()
                TexDistributionError.INVALID_DISTRIBUTION ->
                    """Your installed TeX distribution seems to be corrupt.
                        |Please reinstall your distribution.
                    """.trimMargin()
            }

            client.showMessage(MessageParams(MessageType.Error, message))
            LatexResolver.empty()
        }
    }

    private val componentDatabase: Deferred<LatexComponentDatabase> = async(start = CoroutineStart.LAZY) {
        LatexComponentDatabase.loadOrCreate(
                LatexLanguageServerConfig.COMPONENT_DATABASE_FILE.toFile(),
                resolver.await(),
                progressListener)
    }

    private val completionProvider = CompletionProvider(resolver, componentDatabase)
    private val hoverProvider = HoverProvider(componentDatabase)
    private val diagnosticsProvider = DiagnosticsProvider()

    @JsonDelegate
    override fun getTextDocumentService(): LatexTextDocumentService = this

    override fun getWorkspaceService(): WorkspaceService = this

    fun connect(client: LatexLanguageClient) {
        this.client = client

        launch {
            while (true) {
                workspaceActor.withWorkspace { workspace ->
                    workspace.documentsByUri.values
                            .map { workspace.relatedDocuments(it.uri) }
                            .forEach { componentDatabase.await().getRelatedComponents(it) }
                }

                delay(1000)
            }
        }
    }

    override fun initialize(params: InitializeParams): CompletableFuture<InitializeResult> = future {
        if (params.rootUri != null && params.rootUri.startsWith("file")) {
            val root = URIHelper.parse(params.rootUri)
            loadWorkspace(root)
            workspaceRootDirectory = Paths.get(root)
        }

        val capabilities = ServerCapabilities().apply {
            val syncOptions = TextDocumentSyncOptions().apply {
                openClose = true
                save = SaveOptions(true)
                change = TextDocumentSyncKind.Full
            }
            textDocumentSync = Either.forRight(syncOptions)
            documentSymbolProvider = true
            renameProvider = Either.forLeft(true)
            documentLinkProvider = DocumentLinkOptions(false)
            completionProvider = CompletionOptions(true, listOf("\\", "{", "}", "@", "/"))
            foldingRangeProvider = Either.forLeft(true)
            definitionProvider = true
            hoverProvider = true
            documentFormattingProvider = true
            referencesProvider = true
            documentHighlightProvider = true
        }

        launch {
            resolveIncludes()
        }

        InitializeResult(capabilities)
    }

    override fun initialized(params: InitializedParams?) {
        val watcher = FileSystemWatcher("**/*.log", WatchKind.Create or WatchKind.Change)
        val options = DidChangeWatchedFilesRegistrationOptions(listOf(watcher))
        val registration = Registration("log-watcher", "workspace/didChangeWatchedFiles", options)
        client.registerCapability(RegistrationParams(listOf(registration)))
    }

    override fun didChangeWatchedFiles(params: DidChangeWatchedFilesParams) {
        launch {
            for (change in params.changes) {
                val logPath = File(URIHelper.parse(change.uri)).toPath()
                val texPath = logPath.resolveSibling(logPath.toFile().nameWithoutExtension + ".tex")
                val texUri = texPath.toUri()

                workspaceActor.withWorkspace { workspace ->
                    val document = workspace.documentsByUri[texUri]
                    if (document != null) {
                        try {
                            val log = withContext(Dispatchers.IO) {
                                Files.readAllBytes(logPath).toString(Charsets.UTF_8)
                            }
                            val allErrors = BuildErrorParser.parse(texUri, log)

                            diagnosticsProvider.buildProvider.diagnosticsByUri = allErrors
                                    .groupBy { it.uri }
                                    .mapValues { errors -> errors.value.map { it.toDiagnostic() } }

                            workspace.documentsByUri.values.forEach { publishDiagnostics(it.uri) }
                        } catch (e: IOException) {
                            // File is still locked
                        }
                    }
                }
            }
        }
    }

    override fun didChangeConfiguration(params: DidChangeConfigurationParams) {
    }

    override fun didOpen(params: DidOpenTextDocumentParams) {
        params.textDocument.apply {
            val language = getLanguageById(languageId) ?: return
            val uri = URIHelper.parse(uri)
            workspaceActor.put { Document.create(uri, text, language) }

            launch {
                runLinter(uri, text)
                publishDiagnostics(uri)
                resolveIncludes()
            }
        }
    }

    override fun didChange(params: DidChangeTextDocumentParams) {
        assert(params.contentChanges.size == 1)
        val uri = URIHelper.parse(params.textDocument.uri)
        workspaceActor.put { workspace ->
            val oldDocument = workspace.documentsByUri.getValue(uri)
            val text = params.contentChanges[0].text
            val tree = when (oldDocument) {
                is LatexDocument -> LatexSyntaxTree(text)
                is BibtexDocument -> BibtexSyntaxTree(text)
            }
            oldDocument.copy(text, tree)
        }

        launch {
            publishDiagnostics(uri)
            resolveIncludes()
        }
    }

    override fun didSave(params: DidSaveTextDocumentParams) {
        launch {
            val uri = URIHelper.parse(params.textDocument.uri)
            runLinter(uri, params.text)
            publishDiagnostics(uri)

            val config = client.configuration<BuildConfig>("latex.build", uri)
            if (config.onSave) {
                workspaceActor.withWorkspace { workspace ->
                    val document = workspace.findParent(uri)
                    val identifier = TextDocumentIdentifier(document.uri.toString())
                    build(BuildParams(identifier)).await()
                }
            }
        }
    }

    override fun didClose(params: DidCloseTextDocumentParams) {
    }

    override fun documentSymbol(params: DocumentSymbolParams)
            : CompletableFuture<List<Either<SymbolInformation, DocumentSymbol>>> = future {
        runFeature(SymbolProvider, params.textDocument, params)
                .map { Either.forRight<SymbolInformation, DocumentSymbol>(it) }
    }

    override fun rename(params: RenameParams): CompletableFuture<WorkspaceEdit?> = future {
        runFeature(RenameProvider, params.textDocument, params)
    }

    override fun documentLink(params: DocumentLinkParams)
            : CompletableFuture<List<DocumentLink>> = future {
        runFeature(LatexIncludeLinkProvider, params.textDocument, params)
    }

    override fun completion(params: CompletionParams)
            : CompletableFuture<Either<List<CompletionItem>, CompletionList>> = future {
        val items = runFeature(completionProvider, params.textDocument, params)

        val allIncludes = items.all {
            it.kind == CompletionItemKind.Folder || it.kind == CompletionItemKind.File
        }
        val isIncomplete = !allIncludes || items.size > LatexLanguageServerConfig.COMPLETION_LIMIT
        val list = CompletionList(isIncomplete, items)
        Either.forRight<List<CompletionItem>, CompletionList>(list)
    }

    override fun resolveCompletionItem(unresolved: CompletionItem)
            : CompletableFuture<CompletionItem> = future {
        suspend fun resolveFromMetadataProvider(provider: MetadataProvider) {
            val metadata = provider.getMetadata(unresolved.label)
            unresolved.detail = metadata?.detail
            unresolved.setDocumentation(metadata?.documentation)
        }

        when (unresolved.kind) {
            CompletionItemKind.Constant -> {
                val entry = unresolved.data as JsonPrimitive
                val citation = BibtexCitationActor.cite(entry.asString)

                unresolved.setDocumentation(MarkupContent().apply {
                    kind = MarkupKind.MARKDOWN
                    value = citation
                })
            }
            CompletionItemKind.Class -> {
                resolveFromMetadataProvider(LatexComponentMetadataProvider)
            }
            CompletionItemKind.Interface -> {
                resolveFromMetadataProvider(BibtexEntryTypeMetadataProvider)
            }
            else -> {
            }
        }

        unresolved
    }

    override fun foldingRange(params: FoldingRangeRequestParams)
            : CompletableFuture<List<FoldingRange>> = future {
        runFeature(FoldingProvider, params.textDocument, params)
    }

    override fun definition(params: TextDocumentPositionParams)
            : CompletableFuture<Either<List<Location>, List<LocationLink>>> = future {
        runFeature(DefinitionProvider, params.textDocument, params)
                .let { Either.forLeft<List<Location>, List<LocationLink>>(it) }
    }

    override fun hover(params: TextDocumentPositionParams)
            : CompletableFuture<Hover?> = future {
        runFeature(hoverProvider, params.textDocument, params)
    }

    override fun formatting(params: DocumentFormattingParams)
            : CompletableFuture<List<TextEdit>?> = future {
        val uri = URIHelper.parse(params.textDocument.uri)
        val config = client.configuration<BibtexFormatterConfig>("bibtex.formatting", uri)
        workspaceActor.withWorkspace { workspace ->
            val document = workspace.documentsByUri[uri] as? BibtexDocument
                    ?: return@withWorkspace null
            val formatter =
                    BibtexFormatter(params.options.isInsertSpaces, params.options.tabSize, config.lineLength)
            document.tree.root.children
                    .filterIsInstance<BibtexDeclarationSyntax>()
                    .map { TextEdit(it.range, formatter.format(it)) }
        }
    }

    override fun references(params: ReferenceParams)
            : CompletableFuture<List<Location>> = future {
        runFeature(ReferenceProvider, params.textDocument, params)
    }

    override fun documentHighlight(params: TextDocumentPositionParams)
            : CompletableFuture<List<DocumentHighlight>> = future {
        runFeature(LatexLabelHighlightProvider, params.textDocument, params)
    }

    override fun build(params: BuildParams): CompletableFuture<BuildResult> = future {
        workspaceActor.withWorkspace { workspace ->
            val childUri = URIHelper.parse(params.textDocument.uri)
            val parent = workspace.findParent(childUri)
            val config = client.configuration<BuildConfig>("latex.build", parent.uri)
            BuildEngine.build(parent.uri, config, progressListener)
        }
    }

    override fun forwardSearch(params: TextDocumentPositionParams)
            : CompletableFuture<ForwardSearchResult> = future {
        workspaceActor.withWorkspace { workspace ->
            val childUri = URIHelper.parse(params.textDocument.uri)
            val parent = workspace.findParent(childUri)
            val config = client.configuration<ForwardSearchConfig>("latex.forwardSearch", parent.uri)
            ForwardSearchTool.search(File(childUri), File(parent.uri), params.position.line, config)
        }
    }

    override fun shutdown(): CompletableFuture<Any?> = future {
        null
    }

    override fun exit() {
    }

    private suspend fun publishDiagnostics(uri: URI) {
        workspaceActor.withWorkspace { workspace ->
            val request = FeatureRequest(uri, workspace, Unit, logger)
            val diagnostics = diagnosticsProvider.get(request)
            val params = PublishDiagnosticsParams(uri.toString(), diagnostics)
            client.publishDiagnostics(params)
        }
    }

    private suspend fun resolveIncludes() {
        workspaceActor.withWorkspace { workspace ->
            for (parent in workspace.documentsByUri.values.filterIsInstance<LatexDocument>()) {
                for (include in parent.tree.includes) {
                    if (workspace.resolveDocument(parent.uri, include.path) != null) {
                        continue
                    }

                    for (target in workspace.resolveLinkTargets(parent.uri, include.path)) {
                        val path = Paths.get(target)
                        val child = Workspace.load(path)
                        if (child != null) {
                            workspaceActor.put { child }
                            break
                        }
                    }
                }
            }
        }
    }

    private suspend fun runLinter(uri: URI, text: String) {
        val config = client.configuration<LatexLinterConfig>("latex.lint", uri)
        if (config.onSave) {
            diagnosticsProvider.latexProvider.update(uri, text)
        } else {
            diagnosticsProvider.latexProvider.clear(uri)
        }
    }

    private suspend fun <T, R> runFeature(provider: FeatureProvider<T, R>,
                                          document: TextDocumentIdentifier,
                                          params: T): R {
        return workspaceActor.withWorkspace { workspace ->
            val uri = URIHelper.parse(document.uri)
            val request = FeatureRequest(uri, workspace, params, logger)
            provider.get(request)
        }
    }

    private suspend fun loadWorkspace(root: URI) {
        if (root.scheme == "file") {
            val matcher = FileSystems.getDefault().getPathMatcher("glob:**.{tex,sty,cls,bib}")

            val files = withContext(Dispatchers.IO) {
                Files.walk(Paths.get(root))
                        .filter { Files.isRegularFile(it) }
                        .filter { matcher.matches(it) }
                        .toList()
            }

            files.forEach { loadWorkspaceFile(it) }
        }
    }

    private suspend fun loadWorkspaceFile(file: Path) {
        val document = Workspace.load(file)
        if (document != null) {
            workspaceActor.put { document }
        }
    }
}

File: src/latexLanguageServer.ts
import * as fs from 'fs';
import * as path from 'path';
import {
  CancellationToken,
  CancellationTokenSource,
  CompletionItem,
  CompletionList,
  CompletionParams,
  Definition,
  DefinitionLink,
  Diagnostic,
  DiagnosticSeverity,
  DidChangeTextDocumentParams,
  DidChangeWatchedFilesNotification,
  DidChangeWatchedFilesParams,
  DidChangeWatchedFilesRegistrationOptions,
  DidOpenTextDocumentParams,
  DidSaveTextDocumentParams,
  DocumentFormattingParams,
  DocumentLink,
  DocumentLinkParams,
  FoldingRange,
  FoldingRangeParams,
  Hover,
  InitializeParams,
  InitializeResult,
  Location,
  Range,
  ReferenceParams,
  RenameParams,
  ServerCapabilities,
  TextDocumentIdentifier,
  TextDocumentPositionParams,
  TextDocumentSyncKind,
  TextEdit,
  WatchKind,
  WorkspaceEdit,
} from 'vscode-languageserver';
import { BuildConfig, BuildProvider } from './build';
import { CompletionProvider } from './completion';
import { CompletionItemKind } from './completion/factory';
import { LatexComponentDatabase } from './completion/latex/data/component';
import { COMPONENT_DATABASE_FILE } from './config';
import { definitonProvider } from './definition';
import { diagnosticsProvider } from './diagnostics';
import { BuildErrorKind, parseLog } from './diagnostics/buildLog';
import { LatexLinterConfig } from './diagnostics/latex';
import { Document } from './document';
import { foldingProvider } from './folding';
import { BibtexFormatter, BibtexFormatterConfig } from './formatting/bibtex';
import { ForwardSearchConfig, forwardSearchProvider } from './forwardSearch';
import { HoverProvider } from './hover';
import { getLanguageById, Language } from './language';
import { LanguageServer } from './languageServer';
import { linkProvider } from './link';
import {
  INVALID_DISTRIBUTION_MESSAGE,
  KPSEWHICH_NOT_FOUND_MESSAGE,
  UNKNOWN_DISTRIBUTION_MESSAGE,
} from './messages';
import { generateBibliography } from './metadata/bibtexEntry';
import { getComponentMetadata } from './metadata/component';
import {
  BuildResult,
  BuildTextDocumentParams,
  BuildTextDocumentRequest,
} from './protocol/build';
import {
  ForwardSearchRequest,
  ForwardSearchResult,
} from './protocol/forwardSearch';
import { FeatureContext, FeatureProvider } from './provider';
import { referenceProvider } from './references';
import { renameProvider } from './rename';
import {
  createResolver,
  TexDistributionError,
  TexDistributionErrorKind,
} from './resolver';
import { BibtexSyntaxKind } from './syntax/bibtex/ast';
import { Uri } from './uri';
import { Workspace } from './workspace';

export class LatexLanguageServer extends LanguageServer {
  private readonly workspace = new Workspace();
  private readonly resolver = createResolver();
  private readonly componentDatabase = LatexComponentDatabase.create(
    COMPONENT_DATABASE_FILE,
    this.resolver,
    this.connection.window,
  );

  private readonly completionProvider = CompletionProvider(
    this.resolver,
    this.componentDatabase,
  );

  private readonly hoverProvider = HoverProvider(this.componentDatabase);

  private readonly buildProvider = BuildProvider(
    this.connection.console,
    this.connection.window,
  );

  constructor() {
    super();

    this.resolver.catch(error => {
      if (error instanceof TexDistributionError) {
        const { window } = this.connection;
        switch (error.kind) {
          case TexDistributionErrorKind.KpsewhichNotFound:
            window.showErrorMessage(KPSEWHICH_NOT_FOUND_MESSAGE);
            break;
          case TexDistributionErrorKind.UnknownDistribution:
            window.showErrorMessage(UNKNOWN_DISTRIBUTION_MESSAGE);
            break;
          case TexDistributionErrorKind.InvalidDistribution:
            window.showErrorMessage(INVALID_DISTRIBUTION_MESSAGE);
            break;
        }
      }
    });

    this.connection.onRequest(
      BuildTextDocumentRequest.type,
      this.build.bind(this),
    );

    this.connection.onRequest(
      ForwardSearchRequest.type,
      this.forwardSearch.bind(this),
    );
  }

  public async initialize({
    rootUri,
  }: InitializeParams): Promise<InitializeResult> {
    if (rootUri) {
      const root = Uri.parse(rootUri);
      if (root.isFile()) {
        await this.workspace.loadDirectory(root);
        await this.workspace.loadIncludes();
      }
    }

    const capabilities: ServerCapabilities = {
      textDocumentSync: {
        change: TextDocumentSyncKind.Full,
        save: { includeText: true },
        openClose: true,
      },
      documentSymbolProvider: true,
      renameProvider: true,
      documentLinkProvider: { resolveProvider: false },
      completionProvider: {
        resolveProvider: true,
        triggerCharacters: ['\\', '{', '}', '@', '/'],
      },
      foldingRangeProvider: true,
      definitionProvider: true,
      hoverProvider: true,
      documentFormattingProvider: true,
      referencesProvider: true,
      documentHighlightProvider: true,
    };

    return { capabilities };
  }

  public async initialized() {
    const options: DidChangeWatchedFilesRegistrationOptions = {
      watchers: [
        {
          globPattern: '**/*.log',
          kind: WatchKind.Create | WatchKind.Change,
        },
      ],
    };
    await this.connection.client.register(
      DidChangeWatchedFilesNotification.type,
      options,
    );
  }

  public async didOpenTextDocument({
    textDocument,
  }: DidOpenTextDocumentParams): Promise<void> {
    const language = getLanguageById(textDocument.languageId);
    if (language === undefined) {
      return;
    }

    const uri = Uri.parse(textDocument.uri);
    const document = Document.create(uri, textDocument.text, language);
    this.workspace.put(document);

    await this.runLinter(uri, textDocument.text);
    this.onDidOpenOrChange();
  }

  public didChangeTextDocument({
    textDocument,
    contentChanges,
  }: DidChangeTextDocumentParams): void {
    const uri = Uri.parse(textDocument.uri);
    const text = contentChanges[0].text;
    const { tree } = this.workspace.documents.find(x => x.uri.equals(uri))!;
    const document = Document.create(uri, text, tree.language);
    this.workspace.put(document);

    this.onDidOpenOrChange();
  }

  public async didSaveTextDocument({
    textDocument,
    text,
  }: DidSaveTextDocumentParams): Promise<void> {
    const uri = Uri.parse(textDocument.uri);
    await this.runLinter(uri, text!);
    await this.publishDiagnostics(uri);

    const config: BuildConfig = await this.connection.workspace.getConfiguration(
      {
        section: 'latex.build',
      },
    );
    if (config.onSave) {
      const parent = this.workspace.findParent(uri);
      if (parent !== undefined) {
        const tokenSource = new CancellationTokenSource();
        await this.build(
          { textDocument: { uri: parent.uri.toString() } },
          tokenSource.token,
        );
        tokenSource.dispose();
      }
    }
  }

  public async didChangeWatchedFiles(params: DidChangeWatchedFilesParams) {
    for (const change of params.changes) {
      const { scheme: logScheme, fsPath: logPath } = Uri.parse(change.uri);
      if (logScheme !== 'file') {
        continue;
      }

      const dirname = path.dirname(logPath);
      const texPath = path.join(
        dirname,
        path.basename(logPath).slice(0, -4) + '.tex',
      );
      const texUri = Uri.file(texPath);
      const document = this.workspace.documents.find(x => x.uri.equals(texUri));
      if (document !== undefined) {
        const diagnosticsByUri = new Map<string, Diagnostic[]>();
        diagnosticsProvider.buildProvider.diagnosticsByUri = diagnosticsByUri;
        try {
          const log = await fs.promises.readFile(logPath);
          const allErrors = parseLog(texUri, log.toString());
          allErrors
            .map(x => x.uri.toString())
            .forEach(x => diagnosticsByUri.set(x, []));

          for (const { uri, kind, message, line } of allErrors) {
            const diagnostics = diagnosticsByUri.get(uri.toString())!;
            const severity =
              kind === BuildErrorKind.Error
                ? DiagnosticSeverity.Error
                : DiagnosticSeverity.Warning;

            diagnostics.push({
              message,
              severity,
              range: Range.create(line || 0, 0, line || 0, 0),
              source: 'LaTeX',
            });
          }
        } catch {
          // File is still locked
        }
      }
    }

    for (const { uri } of this.workspace.documents) {
      await this.publishDiagnostics(uri);
    }
  }

  public async hover(
    params: TextDocumentPositionParams,
  ): Promise<Hover | undefined | null> {
    return this.runProvider(this.hoverProvider, params);
  }

  public async completion(
    params: CompletionParams,
  ): Promise<CompletionItem[] | CompletionList | undefined | null> {
    const items = await this.runProvider(this.completionProvider, params);
    const allIncludes = items.every(
      x =>
        x.kind === CompletionItemKind.Folder ||
        x.kind === CompletionItemKind.File,
    );

    return {
      isIncomplete: !allIncludes,
      items,
    };
  }

  public async completionResolve(
    item: CompletionItem,
  ): Promise<CompletionItem> {
    switch (item.data.kind) {
      case CompletionItemKind.Class:
      case CompletionItemKind.Package:
        const metadata = await getComponentMetadata(item.label);
        if (metadata !== undefined) {
          item.detail = metadata.caption;
          item.documentation = metadata.documentation;
        }
        break;
      case CompletionItemKind.Citation:
        item.documentation = generateBibliography(item.data.entry);
        break;
      default:
        break;
    }
    return item;
  }

  public async definition(
    params: TextDocumentPositionParams,
  ): Promise<Definition | DefinitionLink[] | undefined | null> {
    return this.runProvider(definitonProvider, params);
  }

  public async references(
    params: ReferenceParams,
  ): Promise<Location[] | undefined | null> {
    return this.runProvider(referenceProvider, params);
  }

  public async documentFormatting(
    params: DocumentFormattingParams,
  ): Promise<TextEdit[] | undefined | null> {
    const uri = Uri.parse(params.textDocument.uri);
    const document = this.workspace.documents.find(x => x.uri.equals(uri));
    if (document === undefined || document.tree.language !== Language.Bibtex) {
      return null;
    }

    const { insertSpaces, tabSize } = params.options;
    const config: BibtexFormatterConfig = await this.connection.workspace.getConfiguration(
      {
        section: 'bibtex.formatting',
      },
    );

    const formatter = new BibtexFormatter(
      insertSpaces,
      tabSize,
      config.lineLength,
    );

    const edits: TextEdit[] = [];
    document.tree.root.children.forEach(declaration => {
      switch (declaration.kind) {
        case BibtexSyntaxKind.Preamble:
        case BibtexSyntaxKind.String:
        case BibtexSyntaxKind.Entry:
          edits.push({
            range: declaration.range,
            newText: formatter.format(declaration),
          });
          break;
      }
    });
    return edits;
  }

  public async rename(
    params: RenameParams,
  ): Promise<WorkspaceEdit | undefined | null> {
    return this.runProvider(renameProvider, params);
  }

  public async documentLinks(
    params: DocumentLinkParams,
  ): Promise<DocumentLink[] | undefined | null> {
    return this.runProvider(linkProvider, params);
  }

  public async foldingRanges(
    params: FoldingRangeParams,
  ): Promise<FoldingRange[] | undefined | null> {
    return this.runProvider(foldingProvider, params);
  }

  public async build(
    params: BuildTextDocumentParams,
    cancellationToken: CancellationToken,
  ): Promise<BuildResult> {
    const config: BuildConfig = await this.connection.workspace.getConfiguration(
      {
        section: 'latex.build',
      },
    );

    return this.runProvider(
      this.buildProvider,
      { ...params, ...config },
      cancellationToken,
    );
  }

  public async forwardSearch(
    params: TextDocumentPositionParams,
  ): Promise<ForwardSearchResult> {
    const config: ForwardSearchConfig = await this.connection.workspace.getConfiguration(
      {
        section: 'latex.forwardSearch',
      },
    );

    return this.runProvider(forwardSearchProvider, {
      ...params,
      ...config,
    });
  }

  private async onDidOpenOrChange() {
    await this.workspace.loadIncludes();

    this.workspace.documents
      .map(x => this.workspace.relatedDocuments(x.uri))
      .forEach(documents =>
        this.componentDatabase.then(x => x.relatedComponents(documents)),
      );

    for (const document of this.workspace.documents) {
      await this.publishDiagnostics(document.uri);
    }
  }

  private async runLinter(uri: Uri, text: string) {
    const config: LatexLinterConfig = await this.connection.workspace.getConfiguration(
      {
        section: 'latex.lint',
      },
    );

    if (config.onSave) {
      await diagnosticsProvider.latexProvider.update(uri, text);
    } else {
      diagnosticsProvider.latexProvider.clear(uri);
    }
  }

  private async publishDiagnostics(uri: Uri) {
    const params = { textDocument: { uri: uri.toString() } };
    const diagnostics = await this.runProvider(diagnosticsProvider, params);
    await this.connection.sendDiagnostics({
      uri: uri.toString(),
      diagnostics,
    });
  }

  private async runProvider<T, R>(
    provider: FeatureProvider<T, R>,
    params: { textDocument: TextDocumentIdentifier } & T,
    cancellationToken?: CancellationToken,
  ): Promise<R> {
    const uri = Uri.parse(params.textDocument.uri);
    const context = new FeatureContext(uri, this.workspace, params);
    return provider.execute(context, cancellationToken);
  }
}
